executor/src/witgen/jit/symbolic_expression.rs

```diff
@@ -0,0 +1,319 @@
+use std::{
+    fmt::{self, Display, Formatter},
+    ops::{Add, BitAnd, BitOr, Mul, Neg},
+    rc::Rc,
+};
+
+use powdr_number::FieldElement;
+
+use crate::witgen::range_constraints::RangeConstraint;
+
+/// A value that is known at run-time, defined through a complex expression
+/// involving known cells or variables and compile-time constants.
+/// Each of the sub-expressions can have its own range constraint.
+#[derive(Debug, Clone)]
+pub enum SymbolicExpression<T: FieldElement, V> {
+    /// A concrete constant value known at compile time.
+    Concrete(T),
+    /// A symbolic value known at run-time, referencing either a cell or a local variable.
+    Variable(V, Option<RangeConstraint<T>>),
+    BinaryOperation(
+        Rc<Self>,
+        BinaryOperator,
+        Rc<Self>,
+        Option<RangeConstraint<T>>,
+    ),
+    UnaryOperation(UnaryOperator, Rc<Self>, Option<RangeConstraint<T>>),
+}
+
+#[derive(Debug, Clone)]
+pub enum BinaryOperator {
+    Add,
+    Sub,
+    Mul,
+    /// Finite field division.
+    Div,
+    /// Integer division, i.e. convert field elements to unsigned integer and divide.
+    IntegerDiv,
+    BitAnd,
+    BitOr,
+}
+
+#[derive(Debug, Clone)]
+pub enum UnaryOperator {
+    Neg,
+}
+
+impl<T: FieldElement, V> SymbolicExpression<T, V> {
+    pub fn from_var(name: V) -> Self {
+        SymbolicExpression::Variable(name, None)
+    }
+
+    pub fn is_known_zero(&self) -> bool {
+        self.try_to_number().map_or(false, |n| n.is_zero())
+    }
+
+    pub fn is_known_one(&self) -> bool {
+        self.try_to_number().map_or(false, |n| n.is_one())
+    }
+
+    pub fn is_known_minus_one(&self) -> bool {
+        self.try_to_number().map_or(false, |n| n == -T::from(1))
+    }
+
+    pub fn range_constraint(&self) -> Option<RangeConstraint<T>> {
+        match self {
+            SymbolicExpression::Concrete(v) => Some(RangeConstraint::from_value(*v)),
+            SymbolicExpression::Variable(.., rc)
+            | SymbolicExpression::BinaryOperation(.., rc)
+            | SymbolicExpression::UnaryOperation(.., rc) => rc.clone(),
+        }
+    }
```

Comment:
I don't understand yet what we need the range constraints for if we know the value at run-time! As far as I can see, we never set it for `Variable`, and it is only really used once in `AffineSymbolicExpression::transfer_constraints` to get the range constraint of the offset. In the test, the offset is a concrete value.

So, I'm not sure if we need it at all, but if we do, I'd expect that this function receives a `&impl RangeConstraintSet<T>` and recursively computes the constraint, rather than having to store it in the variables & operations!

Reply:
Oh that's a good point, we'll need a constructor for variables where the range constraint is known. This will then be done from the outer interface that also knows the global range constraints etc.

I think it's more convenient to just compute it along.

Reply:
Oh now I know what you mean, I mixed the old impl with a new one and now it's confusing. let me fix.